runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/SamzaMetricsContainer.java

```diff
@@ -53,24 +59,43 @@ public class SamzaMetricsContainer {
 
   private static final Logger LOG = LoggerFactory.getLogger(SamzaMetricsContainer.class);
   private static final String BEAM_METRICS_GROUP = "BeamMetrics";
-  // global metrics container is the default container that can be used in user threads
   public static final String GLOBAL_CONTAINER_STEP_NAME = "GLOBAL_METRICS";
   public static final String USE_SHORT_METRIC_NAMES_CONFIG =
       "beam.samza.metrics.useShortMetricNames";
   public static final String COMMIT_ALL_METRIC_UPDATES =
       "beam.samza.metrics.commitAllMetricUpdates";
+  public static final String DEFER_TO_EXECUTOR_CONFIG = "beam.samza.metrics.deferToExecutor";
+  public static final String METRIC_UPDATE_INTERVAL_SEC_CONFIG =
+      "beam.samza.metrics.updateIntervalSec";
 
   private final MetricsContainerStepMap metricsContainers = new MetricsContainerStepMap();
   private final MetricsRegistryMap metricsRegistry;
   private final boolean useShortMetricNames;
   private final boolean commitAllMetricUpdates;
+  private final boolean deferToExecutor;
+  private final Set<String> activeStepNames = ConcurrentHashMap.newKeySet();
+  // technically this doesn't need to be volatile, but needed to pass spotbugs check
+  private volatile boolean executorServiceStarted = false;
+  private ScheduledExecutorService executorService;
+  private ScheduledFuture<?> scheduledFuture;
+  private final long metricUpdateIntervalSec;
 
   public SamzaMetricsContainer(MetricsRegistryMap metricsRegistry, Config config) {
     this.metricsRegistry = metricsRegistry;
     this.useShortMetricNames = config.getBoolean(USE_SHORT_METRIC_NAMES_CONFIG, false);
     this.commitAllMetricUpdates = config.getBoolean(COMMIT_ALL_METRIC_UPDATES, false);
+    this.deferToExecutor = config.getBoolean(DEFER_TO_EXECUTOR_CONFIG, false);
+    metricUpdateIntervalSec = config.getLong(METRIC_UPDATE_INTERVAL_SEC_CONFIG, 1L);
     this.metricsRegistry.metrics().put(BEAM_METRICS_GROUP, new ConcurrentHashMap<>());
-    LOG.info("Creating Samza metrics container with userShortMetricName = {}", useShortMetricNames);
+
+    LOG.info(
+        "Creating Samza metrics container with deferToExecutor={}, metricUpdateIntervalSec={}, useShortMetricNames={}, commitAllMetricUpdates={}",
+        deferToExecutor,
+        metricUpdateIntervalSec,
+        useShortMetricNames,
+        commitAllMetricUpdates);
+    // Register a shutdown hook to gracefully shutdown the executor service
+    Runtime.getRuntime().addShutdownHook(new Thread(this::shutdownExecutorService));
```

Comment:
What will happen in the case where `beam.samza.metrics.deferToExecutor` is set to false, but the shutdownExecutorService  hook is added?

Reply:
This line should handle it
```
    if (scheduler != null && !scheduler.isShutdown()) {
```
scheduler would be null and so shutdownExecutorService would NOOP.

Reply:
Although this PR is experimental, we will be releasing a Beam version that will be available to other Flink jobs. Can we ensure that the job’s behavior remains the same as before when `beam.samza.metrics.deferToExecutor` is set to false? Therefore, can we move `Runtime.getRuntime().addShutdownHook ` inside the `if (deferToExecutor)` block?